kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -303,4 +428,166 @@ private static boolean hasPhysicalName(StructField field) {
     }
+
+    private static boolean hasNestedColumnIds(StructField field) {
+        return field.getMetadata().contains(COLUMN_MAPPING_NESTED_IDS_KEY);
+    }
+
+    private static FieldMetadata getNestedColumnIds(StructField field) {
+        return field.getMetadata().getMetadata(COLUMN_MAPPING_NESTED_IDS_KEY);
+    }
+
+    private static int getMaxNestedColumnId(StructField field) {
+        return getNestedColumnIds(field).getEntries().values().stream()
+                .filter(Long.class::isInstance)
+                .map(Long.class::cast)
+                .max(Comparator.naturalOrder())
+                .orElse(0L)
+                .intValue();
+    }
+
+    /**
+     * Adds the nested field IDs required by Iceberg.
+     *
+     * In parquet, list-type columns have a nested, implicitly defined {@code element} field and
+     * map-type columns have implicitly defined {@code key} and {@code value} fields. By default,
+     * Spark does not write field IDs for these fields in the parquet files. However, Iceberg
+     * requires these *nested* field IDs to be present. This method rewrites the specified
+     * schema to add those nested field IDs.
+     *
+     * Nested field IDs are stored in a map as part of the metadata of the
+     * *nearest* parent {@link StructField}.
+     * For example, consider the following schema:
+     *
+     * col1 ARRAY(INT)
+     * col2 MAP(INT, INT)
+     * col3 STRUCT(a INT, b ARRAY(STRUCT(c INT, d MAP(INT, INT))))
+     *
+     * col1 is a list and so requires one nested field ID for the {@code element} field in parquet.
+     * This nested field ID will be stored in a map that is part
+     * of col1's {@link StructField#getMetadata()}.
+     * The same applies to the nested field IDs for col2's implicit {@code key} and {@code value}
+     * fields. col3 itself is a Struct, consisting of an integer field and a list field named 'b'.
+     * The nested field ID for the list of 'b' is stored in b's {@link StructField#getMetadata()}.
+     * Finally, the list type itself is again a struct consisting of an integer field
+     * and a map field named 'd'.
+     * The nested field IDs for the map of 'd' are stored in d's {@link StructField#getMetadata()}.
+     *
+     * @param schema  The schema to which nested field IDs should be added
+     * @param startId The first field ID to use for the nested field IDs
+     */
+    private static StructType rewriteFieldIdsForIceberg(StructType schema, AtomicInteger startId) {
+        StructType newSchema = new StructType();
+        for (StructField field : schema.fields()) {
+            newSchema = newSchema.add(
+                    transformSchema(startId, field, "" /** current column path */));
+        }
+        return newSchema;
+    }
+
+    /**
+     * Recursively visits each nested struct / array / map type and returns a new
+     * {@link StructField} with updated {@link FieldMetadata}. For array / map types that
```

Comment:
I think this would be clearer:

`For array / map types, the field IDs of their nested elements are under the {@link ColumnMapping#COLUMN_MAPPING_NESTED_IDS_KEY} key.`